File: wallet-kit/src/main/java/bitcoin/walllet/kit/network/message/InvMessage.java
package bitcoin.walllet.kit.network.message;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.util.Arrays;

import bitcoin.walllet.kit.common.io.BitcoinInput;
import bitcoin.walllet.kit.common.io.BitcoinOutput;
import bitcoin.walllet.kit.common.util.HashUtils;
import bitcoin.walllet.kit.struct.InvVect;

public class InvMessage extends Message {

    InvVect[] inventory;

    public InvMessage() {
        super("inv");
        inventory = new InvVect[0];
    }

    public InvMessage(byte[] payload) throws IOException {
        super("inv");
        try (BitcoinInput input = new BitcoinInput(new ByteArrayInputStream(payload))) {
            long count = input.readVarInt(); // do not store count
            this.inventory = new InvVect[(int) count];
            for (int i = 0; i < this.inventory.length; i++) {
                this.inventory[i] = new InvVect(input);
            }
        }
    }

    public byte[][] getBlockHashes() {
        return Arrays.stream(this.inventory).filter((iv) -> {
            return iv.type == InvVect.MSG_BLOCK;
        }).map((iv) -> {
            return iv.hash;
        }).toArray(byte[][]::new);
    }

    public String[] getBlockHashesAsString() {
        return Arrays.stream(this.inventory).filter((iv) -> {
            return iv.type == InvVect.MSG_BLOCK;
        }).map((iv) -> {
            return HashUtils.toHexStringAsLittleEndian(iv.hash);
        }).toArray(String[]::new);
    }

    @Override
    protected byte[] getPayload() {
        BitcoinOutput output = new BitcoinOutput();
        output.writeVarInt(this.inventory.length);
        for (int i = 0; i < this.inventory.length; i++) {
            output.write(this.inventory[i].toByteArray());
        }
        return output.toByteArray();
    }

    @Override
    public String toString() {
        return "InvMessage(" + this.inventory.length + ": ["
                + String.join(", ", Arrays.stream(this.inventory).map((inv) -> {
            return inv.type + ":" + HashUtils.toHexStringAsLittleEndian(inv.hash);
        }).limit(10).toArray(String[]::new)) + "])";
    }
}

File: wallet-kit/src/main/kotlin/bitcoin/wallet/kit/messages/InvMessage.kt
package bitcoin.wallet.kit.messages

import bitcoin.wallet.kit.models.InventoryItem
import bitcoin.walllet.kit.io.BitcoinInput
import bitcoin.walllet.kit.io.BitcoinOutput
import bitcoin.walllet.kit.utils.HashUtils
import java.io.ByteArrayInputStream
import java.io.IOException

/**
 * Inventory Message
 *
 *  Size        Field           Description
 *  ====        =====           ===========
 *  VarInt      Count           Number of inventory items
 *  Variable    InventoryItem   One or more inventory items
 */
class InvMessage : Message {

    lateinit var inventory: Array<InventoryItem>

    constructor() : super("inv") {
        inventory = arrayOf()
    }

    constructor(type: Int, hash: ByteArray) : super("inv") {
        val inv = InventoryItem()
        inv.type = type
        inv.hash = hash
        inventory = arrayOf(inv)
    }

    @Throws(IOException::class)
    constructor(payload: ByteArray) : super("inv") {
        BitcoinInput(ByteArrayInputStream(payload)).use { input ->
            val count = input.readVarInt() // do not store count
            inventory = Array(count.toInt()) {
                InventoryItem(input)
            }
        }
    }

    fun getBlockHashes(): Array<ByteArray> {
        return inventory
                .filter { iv -> iv.type == InventoryItem.MSG_BLOCK }
                .map { iv -> iv.hash }
                .toTypedArray()
    }

    fun getBlockHashesAsString(): Array<String> {
        return inventory
                .filter { iv -> iv.type == InventoryItem.MSG_BLOCK }
                .map { iv -> HashUtils.toHexStringAsLittleEndian(iv.hash) }
                .toTypedArray()
    }

    override fun getPayload(): ByteArray {
        val output = BitcoinOutput()
        output.writeVarInt(inventory.size.toLong())
        for (i in inventory.indices) {
            output.write(inventory[i].toByteArray())
        }
        return output.toByteArray()
    }

    override fun toString(): String {
        val invList = inventory.take(10)
                .map { inv -> inv.type.toString() + ":" + HashUtils.toHexStringAsLittleEndian(inv.hash) }
                .toTypedArray()
                .joinToString()

        return ("InvMessage(" + inventory.size + ": [" + invList + "])")
    }
}
